File: algorithms/2_medium/W/00916_word_subsets/solution.c
#define FREQS_LEN 26

static inline uint8_t max_u8(const uint8_t a, const uint8_t b)
{
    return a > b ? a : b;
}

static inline void calc_freqs(const char* s, uint8_t* res)
{
    for (size_t i = 0; s[i] != '\0'; i += 1) {
        const char ch = s[i];
        assert(ch >= 'a' && ch <= 'z' && res[ch - 'a'] < UINT8_MAX);
        res[ch - 'a'] += 1;
    }
}
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
char** wordSubsets(char** words1, int len1, char** words2, const int len2, int* return_len)
{
    assert(len1 > 0 && len2 > 0 && return_len != NULL && words1 != NULL && words2 != NULL);
    char** res = NULL;
    size_t res_cap = 0;
    size_t res_len = 0;
    uint8_t max_freqs2[FREQS_LEN] = { 0 };
    for (int i = 0; i < len2; i += 1) {
        uint8_t freqs[FREQS_LEN] = { 0 };
        calc_freqs(words2[i], freqs);
        for (int i = 0; i < FREQS_LEN; i += 1) {
            max_freqs2[i] = max_u8(max_freqs2[i], freqs[i]);
        }
    }

    for (int i = 0; i < len1; i += 1) {
        uint8_t freqs[FREQS_LEN] = { 0 };
        calc_freqs(words1[i], freqs);
        bool is_valid = true;

        for (int i = 0; i < FREQS_LEN; i += 1) {
            if (max_freqs2[i] > freqs[i]) {
                is_valid = false;
                break;
            }
        }

        if (is_valid) {
            if (res_cap == res_len) {
                res_cap = res_cap == 0 ? 4 : 2 * res_cap;
                res = realloc(res, res_cap * sizeof(*res));
                assert(res != NULL);
            }
            res[res_len] = strdup(words1[i]);
            assert(res[res_len] != NULL);
            res_len += 1;
        }
    }

    assert(res_len <= (size_t)INT_MAX);
    *return_len = (int)res_len;
    return res;
}

File: algorithms/2_medium/W/00916_word_subsets/solution.rs
impl Solution {
    const FREQS_LEN: usize = 26;

    #[inline(always)]
    fn calc_freqs(bytes: &[u8]) -> [u8; Self::FREQS_LEN] {
        assert!(bytes.len() <= u8::MAX as usize);
        let mut res = [0_u8; Self::FREQS_LEN];
        for &b in bytes {
            assert!(b >= b'a' && b <= b'z');
            res[(b - b'a') as usize] += 1;
        }
        res
    }

    pub fn word_subsets(mut words1: Vec<String>, words2: Vec<String>) -> Vec<String> {
        let mut max_freqs2 = [0_u8; Self::FREQS_LEN];
        for w in words2 {
            let freqs = Self::calc_freqs(w.as_bytes());
            for i in 0..Self::FREQS_LEN {
                max_freqs2[i] = max_freqs2[i].max(freqs[i]);
            }
        }
        let mut res: Vec<String> = Vec::new();

        for w in words1{
            let freqs = Self::calc_freqs(w.as_bytes());
            let mut is_valid = true;

            for i in 0..Self::FREQS_LEN {
                if max_freqs2[i] > freqs[i] {
                    is_valid = false;
                    break;
                }
            }

            if is_valid {
                res.push(w);
            }
        }

        res
    }
}
